### PR TITLE
#803 - Hotfix - resolve unbound variable and exception handling

### PR DESCRIPTION
**Pull request by ChrisJohnson-CDJ**
# Description

Paired with Kyle to trace the error/exception paths.
Complexity metric increased.
Iterated the changes, updating failing unit tests.

1. Address errors resulting in exception handler throwing UnboundLocal
2. Ensure sample_notification_history has a valid reference, uuid4
3. Moved notification lookup out to main celery task to ensure notification available to exception handlers
4. Added KeyError handler to handle malformed payloads
5. Updated unittests for change function signatures
6. Added coverage for missing/None reference but update to sample_notification_history should fix this in tests.
7. Added test for notification not found (forcing no retry to catch error message)

issue #803 

## How Has This Been Tested?

Unit testing and DEV deploy regression tests.

image

image

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an [appropriate title](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/team_agreement.md#naming-prs): #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/pull_request_labels.md#vanotify-labels) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [ ] My code follows the [style guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/style_guide.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works. [Testing guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/testing_guide.md)
- [x] I have ensured the latest main is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket was moved into the DEV test column when I began testing this change

### CODE REVIEW

File: app/celery/process_ses_receipts_tasks.py [hunk 1]
```diff
@@ -36,31 +35,52 @@
...
 @statsd(namespace='tasks')
-def process_ses_results(  # noqa: C901 (too complex 14 > 10)
+def process_ses_results(  # noqa: C901 (too complex 19 > 10)
     self,
     response,
 ):
     current_app.logger.debug('Full SES result response: %s', response)
     try:
         ses_message = json.loads(response['Message'])
+
```
Comment (ChrisJohnson-CDJ) [inline app/celery/process_ses_receipts_tasks.py:45]: reference checks moved to common early guard

File: app/celery/process_ses_receipts_tasks.py [hunk 2]
```diff
@@ -36,31 +35,52 @@
...
+        reference = ses_message['mail']['messageId']
+        if not reference:
+            current_app.logger.warning(
+                'SES complaint: unable to lookup notification, messageId (reference) was None | ses_message: %s',
+                ses_message,
+            )
+            return
+
         notification_type = ses_message.get('eventType')
```
Comment (kalbfled) [inline app/celery/process_ses_receipts_tasks.py:54]: None of this needs to be in the "try" block.
```suggestion
    ses_message = json.loads(response.get('Message', {}))

    reference = ses_message.get('mail', {}).get('messageId')
    if not reference:
        current_app.logger.warning(
            'SES complaint: unable to lookup notification, messageId (reference) was None | ses_message: %s',
            ses_message,
        )
        return

    notification_type = ses_message.get('eventType')

    try:
```

File: app/celery/process_ses_receipts_tasks.py [hunk 3]
```diff
@@ -36,31 +35,52 @@
...
+                message_time = iso8601.parse_date(ses_message['mail']['timestamp']).replace(tzinfo=None)
+                if datetime.utcnow() - message_time < timedelta(minutes=5):
+                    self.retry(queue=QueueNames.RETRY)
+                else:
+                    current_app.logger.warning('SES complaint: notification not found | reference: %s', reference)
+                return
+
+            complaint, recipient_email = handle_ses_complaint(ses_message, notification)
+            publish_complaint(complaint, notification, recipient_email)
+            return
```
Comment (kalbfled) [inline app/celery/process_ses_receipts_tasks.py:74]: You need an "else" here because notification_type could be None.  Maybe that shouldn't happen, but it's possible according to the code.  Another option would be to `assert notification_type is not None`.  I like to use assertions to codify assumptions.

Yet another option would be to test above `if notification_type is None` and early-out accordingly.

File: app/celery/process_ses_receipts_tasks.py [hunk 4]
```diff
@@ -36,31 +35,52 @@
...
+                'SES complaint: unable to lookup notification, messageId (reference) was None | ses_message: %s',
+                ses_message,
+            )
+            return
+
         notification_type = ses_message.get('eventType')
 
         if notification_type == 'Bounce':
             # Bounces have ran their course with AWS and should be considered done. Clients can retry soft bounces.
             notification_type = determine_notification_bounce_type(notification_type, ses_message)
```
Comment (kalbfled) [inline app/celery/process_ses_receipts_tasks.py:58]: Can this raise an exception?  If not, most of this code shouldn't be an a "try" block either.

File: app/celery/process_ses_receipts_tasks.py [hunk 5]
```diff
@@ -152,6 +172,9 @@ def process_ses_results(  # noqa: C901 (too complex 14 > 10)
     except JSONDecodeError:
         current_app.logger.exception('Error decoding SES results: full response data: %s', response)
 
+    except KeyError:
+        current_app.logger.exception('AWS message malformed: full response data: %s', response)
```
Comment (kalbfled) [inline app/celery/process_ses_receipts_tasks.py:176]: This "except" isn't necessary using to code I suggested above.

File: app/notifications/notifications_ses_callback.py [hunk 6]
```diff
@@ -16,19 +16,13 @@ def determine_notification_bounce_type(
...
     )
-    try:
-        reference = ses_message['mail']['messageId']
-    except KeyError:
-        current_app.logger.exception('Complaint from SES failed to get reference from message.')
-        return
-    notification = dao_get_notification_history_by_reference(reference)
-    ses_complaint = ses_message.get('complaint', None)
 
+    ses_complaint = ses_message.get('complaint', None)
```
Comment (kalbfled) [inline app/notifications/notifications_ses_callback.py:25]:
```suggestion
    ses_complaint = ses_message.get('complaint')
```

File: tests/app/conftest.py [hunk 7]
```diff
@@ -1555,7 +1555,7 @@ def _sample_notification_history(
             key_type=key_type,
             api_key=api_key,
             api_key_id=api_key.id,
-            reference=reference,
+            reference=reference or uuid4(),
```
Comment (kalbfled) [inline tests/app/conftest.py:1558]: This looks like JavaScript.  Please use the Pythonic syntax, which is admittedly more verbose in this case.
```suggestion
            reference=reference if (reference is not None) else uuid4(),
```
Are you sure this shouldn't be `str(uuid4())`?

File: tests/app/notifications/test_notifications_ses_callback.py [hunk 8]
```diff
@@ -30,35 +30,24 @@ def test_process_ses_results_in_complaint(
     sample_notification,
 ):
     ref = str(uuid4())
-    notification = sample_notification(gen_type=EMAIL_TYPE, reference=ref)
+    notification: Notification = sample_notification(gen_type=EMAIL_TYPE, reference=ref)
```
Comment (kalbfled) [inline tests/app/notifications/test_notifications_ses_callback.py:33]: The annotation isn't helpful here.  The type is obvious.
```suggestion
    notification = sample_notification(gen_type=EMAIL_TYPE, reference=ref)
```

**Review by kalbfled — CHANGES_REQUESTED** [review]

File: tests/app/conftest.py (hunk 7 above)
Comment (ChrisJohnson-CDJ) [reply tests/app/conftest.py:1558]: The field in the model is a db.string but it doesn't hurt to be safe.

File: app/celery/process_ses_receipts_tasks.py (hunk 5 above)
Comment (ChrisJohnson-CDJ) [reply app/celery/process_ses_receipts_tasks.py:176]: There are still many place in the code that could throw KeyError due to a malformed SES payload. A refactor should perform more thorough validation in one place.

File: app/celery/process_ses_receipts_tasks.py (hunk 4 above)
Comment (ChrisJohnson-CDJ) [reply app/celery/process_ses_receipts_tasks.py:58]: Same comment as below regarding KeyErrors due to a malformed payload.

**Review by EvanParish — APPROVED** [review]

**Review by kalbfled — APPROVED** [review]

**Review by cris-oddball — CHANGES_REQUESTED** [review]
please use the correct label. hotfix is not pone of major, minor, patch

**Review by cris-oddball — APPROVED** [review]
